File: src/main/java/com/kobe/warehouse/service/report/CustomerSegmentationReportService.java
package com.kobe.warehouse.service.report;

import com.kobe.warehouse.service.dto.report.CustomerSegmentationDTO;
import java.util.List;
import java.util.Map;

public interface CustomerSegmentationReportService {
    /**
     * Get all customer segmentation data using RFM analysis
     *
     * @return List of customer segmentation records
     */
    List<CustomerSegmentationDTO> getAllCustomerSegmentation();

    /**
     * Get customers filtered by classification
     *
     * @param classification The customer classification to filter by
     * @return List of customers for the classification
     */
    List<CustomerSegmentationDTO> getCustomersByClassification(CustomerSegmentationDTO.CustomerClassification classification);

    /**
     * Get champion customers (highest RFM scores)
     *
     * @return List of champion customers
     */
    List<CustomerSegmentationDTO> getChampionCustomers();

    /**
     * Get at-risk customers (need attention)
     *
     * @return List of at-risk customers
     */
    List<CustomerSegmentationDTO> getAtRiskCustomers();

    /**
     * Get count of customers by classification
     *
     * @return Map of classification to count
     */
    Map<CustomerSegmentationDTO.CustomerClassification, Long> getCustomerCountByClassification();

    /**
     * Get customer segmentation for a specific customer
     *
     * @param customerId The customer ID
     * @return Customer segmentation data
     */
    CustomerSegmentationDTO getCustomerSegmentation(Integer customerId);

    /**
     * Export customer segmentation report as PDF
     *
     * @return PDF bytes
     */
    byte[] exportCustomerSegmentationToPdf();
}

File: pharmaSmart-app/src/main/webapp/app/entities/reports/services/customer-segmentation-report.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient, HttpResponse, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { SERVER_API_URL } from 'app/app.constants';
import { ICustomerSegmentation, CustomerClassification } from 'app/shared/model/report/customer-segmentation.model';

type EntityArrayResponseType = HttpResponse<ICustomerSegmentation[]>;
type EntityResponseType = HttpResponse<ICustomerSegmentation>;
type CountResponseType = HttpResponse<Record<CustomerClassification, number>>;

@Injectable({ providedIn: 'root' })
export class CustomerSegmentationReportService {
  private readonly resourceUrl = SERVER_API_URL + 'api/customers/segmentation';
  private readonly http = inject(HttpClient);

  /**
   * Get all customer segmentation data using RFM analysis
   */
  getAllCustomerSegmentation(): Observable<EntityArrayResponseType> {
    return this.http.get<ICustomerSegmentation[]>(this.resourceUrl, { observe: 'response' });
  }

  /**
   * Get customers filtered by classification
   */
  getCustomersByClassification(classification: CustomerClassification): Observable<EntityArrayResponseType> {
    const params = new HttpParams().set('classification', classification);
    return this.http.get<ICustomerSegmentation[]>(`${this.resourceUrl}/classification`, { params, observe: 'response' });
  }

  /**
   * Get champion customers (highest RFM scores)
   */
  getChampionCustomers(): Observable<EntityArrayResponseType> {
    return this.http.get<ICustomerSegmentation[]>(`${this.resourceUrl}/champions`, { observe: 'response' });
  }

  /**
   * Get at-risk customers (need attention)
   */
  getAtRiskCustomers(): Observable<EntityArrayResponseType> {
    return this.http.get<ICustomerSegmentation[]>(`${this.resourceUrl}/at-risk`, { observe: 'response' });
  }

  /**
   * Get count of customers by classification
   */
  getCustomerCountByClassification(): Observable<CountResponseType> {
    return this.http.get<Record<CustomerClassification, number>>(`${this.resourceUrl}/count`, { observe: 'response' });
  }

  /**
   * Get customer segmentation for a specific customer
   */
  getCustomerSegmentation(customerId: number): Observable<EntityResponseType> {
    return this.http.get<ICustomerSegmentation>(`${this.resourceUrl}/${customerId}`, { observe: 'response' });
  }

  /**
   * Export customer segmentation report as PDF
   */
  exportCustomerSegmentationToPdf(): Observable<HttpResponse<Blob>> {
    return this.http.get(`${this.resourceUrl}/export`, { observe: 'response', responseType: 'blob' });
  }
}
